test/e2e/storage/testsuites/provisioning.go

```diff
@@ -0,0 +1,370 @@
+
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+
+	"k8s.io/api/core/v1"
+	storage "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"k8s.io/kubernetes/test/e2e/storage/drivers"
+	"k8s.io/kubernetes/test/e2e/storage/testpatterns"
+	imageutils "k8s.io/kubernetes/test/utils/image"
+)
+
+// StorageClassTest represents parameters to be used by provisioning tests
+type StorageClassTest struct {
+	Name               string
+	CloudProviders     []string
+	Provisioner        string
+	Parameters         map[string]string
+	DelayBinding       bool
+	ClaimSize          string
+	ExpectedSize       string
+	PvCheck            func(volume *v1.PersistentVolume) error
+	NodeName           string
+	SkipWriteReadCheck bool
+	VolumeMode         *v1.PersistentVolumeMode
+}
+
+type provisioningTestSuite struct {
+	tsInfo TestSuiteInfo
+}
+
+var _ TestSuite = &provisioningTestSuite{}
+
+// InitProvisioningTestSuite returns provisioningTestSuite that implements TestSuite interface
+func InitProvisioningTestSuite() TestSuite {
+	return &provisioningTestSuite{
+		tsInfo: TestSuiteInfo{
+			name: "provisioning",
+			testPatterns: []testpatterns.TestPattern{
+				testpatterns.DefaultFsDynamicPV,
+			},
+		},
+	}
+}
+
+func (p *provisioningTestSuite) getTestSuiteInfo() TestSuiteInfo {
+	return p.tsInfo
+}
+
+func (p *provisioningTestSuite) skipUnsupportedTest(pattern testpatterns.TestPattern, driver drivers.TestDriver) {
+}
+
+func createProvisioningTestInput(pattern testpatterns.TestPattern, resource provisioningTestResource) provisioningTestInput {
+	driver := resource.driver
+	dInfo := driver.GetDriverInfo()
+	f := dInfo.Framework
+
+	input := provisioningTestInput{
+		testCase: StorageClassTest{
+			ClaimSize:    resource.claimSize,
+			ExpectedSize: resource.claimSize,
+		},
+		cs:    f.ClientSet,
+		pvc:   resource.pvc,
+		sc:    resource.sc,
+		dInfo: dInfo,
+	}
+
+	if dInfo.Config.ClientNodeName != "" {
+		input.testCase.NodeName = dInfo.Config.ClientNodeName
+	}
+
+	return input
+}
+
+func (p *provisioningTestSuite) execTest(driver drivers.TestDriver, pattern testpatterns.TestPattern) {
+	Context(getTestNameStr(p, pattern), func() {
+		var (
+			resource     provisioningTestResource
+			input        provisioningTestInput
+			needsCleanup bool
+		)
+
+		BeforeEach(func() {
+			needsCleanup = false
+			// Skip unsupported tests to avoid unnecessary resource initialization
+			skipUnsupportedTest(p, driver, pattern)
+			needsCleanup = true
+
+			// Setup test resource for driver and testpattern
+			resource = provisioningTestResource{}
+			resource.setupResource(driver, pattern)
+
+			// Create test input
+			input = createProvisioningTestInput(pattern, resource)
+		})
+
+		AfterEach(func() {
+			if needsCleanup {
+				resource.cleanupResource(driver, pattern)
+			}
+		})
+
+		testProvisioning(&input)
```

Comment:
why is this a pointer?

Reply:
This is related to  https://github.com/kubernetes/kubernetes/pull/66577#discussion_r210006095 , although implementation has been changed so much since then.

In summary, we are using pattern 2 of "Global Shared Behaviors" described in https://onsi.github.io/ginkgo/#global-shared-behaviors, and it requires all the inputs for It() to be in one struct and the struct is required to be passed as a pointer.

And you are right that it should be commented in somewhere, because it is a bit tricky part and is not obvious.

Reply:
cool, thanks for the explaination! can we just add a comment above this line then